compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3854,6 +3854,176 @@ static void arraySetDefault(TR::Node* node, uint8_t elementSize, TR::Register* a
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(addressReg, 32, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(addressReg, 48, cg), xmmValueReg, cg);
+   generateRegImmInstruction(TR::InstOpCode::ADDRegImms(), node, addressReg, 64, cg);
+   generateRegImmInstruction(TR::InstOpCode::SUBRegImms(), node, sizeReg, 64, cg);
+   generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 64, cg);
+   generateLabelInstruction(TR::InstOpCode::JAE4, node, loopLabel, cg);
+   // Residue
+   generateLabelInstruction(TR::InstOpCode::label, node, residueLabel, cg);
+   // At this point we have less than 64 bytes to store
+   // Calculate the address of the last 3 aligned stores
+   generateRegMemInstruction(TR::InstOpCode::LEARegMem(), node, scratch1Reg, generateX86MemoryReference(addressReg, sizeReg, 0, -48, cg), cg);
+   generateRegImmInstruction(TR::InstOpCode::ANDRegImms(), node, scratch1Reg, -16, cg);
+   // Calculate the address of the last (unaligned) store
+   generateRegMemInstruction(TR::InstOpCode::LEARegMem(), node, scratch2Reg, generateX86MemoryReference(addressReg, sizeReg, 0, -16, cg), cg);
+   // Do the stores
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(scratch1Reg, 0, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(scratch1Reg, 16, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVAPSMemReg, node, generateX86MemoryReference(scratch1Reg, 32, cg), xmmValueReg, cg);
+   generateMemRegInstruction(TR::InstOpCode::MOVUPSMemReg, node, generateX86MemoryReference(scratch2Reg, 0, cg), xmmValueReg, cg);
+   }
+
+static void arraySetXMM(TR::Node* node, uint8_t elementSize, TR::Register* addressReg, TR::Register* valueReg, TR::Register* sizeReg, const uintptr_t *size, TR::CodeGenerator* cg)
+   {
+   // This code is loosely based on the approach discussed in
+   // "Building Faster AMD64 Memset Routines" by Joe Bialek (January 11, 2021).
+   // https://msrc.microsoft.com/blog/2021/01/building-faster-amd64-memset-routines/
+   // We reduce path length and handle unaligned bytes more efficiently
+   // by setting some bytes multiple times, on the assumption
+   // that stores to overlapping memory ranges are cheaper than executing
+   // extra comparisons and branches to set each byte exactly once.
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+
+   TR::Register *scratch1Reg = cg->allocateRegister(TR_GPR);
+   TR::Register *scratch2Reg = cg->allocateRegister(TR_GPR);
+   TR::Register *xmmValueReg = cg->allocateRegister(TR_FPR);
+
+   generateRegRegInstruction(TR::InstOpCode::MOVDRegReg4, node, xmmValueReg, valueReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::VPBROADCASTBRegReg, node, xmmValueReg, xmmValueReg, cg);
```

Comment:
I was working on a [helper](https://github.com/eclipse-omr/omr/commit/0f45a7820b8f462898933745a49cd9180fbcfc32) for broadcast operations, you can contribute it here if you'd like, and modify as needed to use the vpbroadcastb instruction when possible.